File: competitive-programming/leet-code/301/removeInvalidParentheses.js
let removeInvalidParentheses = function (s) {
    const isValid = str => {
        let count = 0;
        for (let ch of str) {
            if (ch === '(') count++;
            else if (ch === ')') count--;
            if (count < 0) return false;
        }

        return count === 0;
    }

    let res = [], visited = new Set([s]), queue = [s];

    while (queue.length) {
        let next = [];
        for (let str of queue) if (isValid(str)) res.push(str);
        if (res.length) return res;
        for (let str of queue) {
            for (let i = 0; i < str.length; i++) {
                if (!'()'.includes(str[i])) continue;
                let t = str.slice(0, i) + str.slice(i + 1);
                if (!visited.has(t)) {
                    visited.add(t);
                    next.push(t);
                }
            }
        }

        queue = next;
    }

    return [""];
}

File: competitive-programming/leet-code/301/remove_invalid_parentheses.py
from typing import List, Set

def remove_invalid_parentheses(s: str) -> List[str]:
    def is_valid(string: str) -> bool:
        count = 0
        for ch in string:
            if ch == '(':
                count += 1
            elif ch == ')':
                count -= 1
            if count < 0:
                return False
        return count == 0

    res: List[str] = []
    visited: Set[str] = {s}
    queue: List[str] = [s]

    while queue:
        next_queue: List[str] = []
        
        for string in queue:
            if is_valid(string):
                res.append(string)
        
        if res:
            return res
        
        for string in queue:
            for i in range(len(string)):
                if string[i] not in '()':
                    continue
        
                new_string = string[:i] + string[i + 1:]
                if new_string not in visited:
                    visited.add(new_string)
                    next_queue.append(new_string)
        
        queue = next_queue

    return [""]